src/key.js

```diff
@@ -513,32 +528,31 @@ Key.prototype.getValidUsers = async function(date=new Date(), allowExpired=false
       return;
     }
     const dataToVerify = { userid: user.userId , key: primaryKey };
-    for (let j = 0; j < user.selfCertifications.length; j++) {
-      const cert = user.selfCertifications[j];
-      // skip if certificate is not the most recent
-      if ((cert.isPrimaryUserID && cert.isPrimaryUserID < lastPrimaryUserID) ||
-          (!lastPrimaryUserID && cert.created < lastCreated)) {
-        continue;
-      }
-      // skip if certificates is invalid, revoked, or expired
-      // eslint-disable-next-line no-await-in-loop
-      if (!(cert.verified || await cert.verify(primaryKey, dataToVerify))) {
-        continue;
-      }
-      // eslint-disable-next-line no-await-in-loop
-      if (cert.revoked || await user.isRevoked(primaryKey, cert, null, date)) {
-        continue;
-      }
-      if (!allowExpired && cert.isExpired(date)) {
-        continue;
-      }
-      lastPrimaryUserID = cert.isPrimaryUserID;
-      lastCreated = cert.created;
-      validUsers.push({ index: i, user: user, selfCertification: cert });
+    const cert = getLatestSignature(user.selfCertifications);
+    // skip if certificate is not the most recent
+    if ((cert.isPrimaryUserID && cert.isPrimaryUserID < lastPrimaryUserID) ||
```

Comment:
I know this code is copied from before, but this check is a vestige of our previous implementation for getPrimaryUser, and I am realizing it is no longer really correct if this function is meant to return *all* valid users. This is especially important as it relates to checking expiration time.

So what I propose that we do is remove the lastCreated and lastPrimaryUserID variables, remove this check, and then make this for-loop a Promise.all.

Reply:
Yeah, that code is incorrect. But the only calling function that uses the whole list of users is `Key#getExpirationTime`, which in turn is never used (although it's part of the public API, of course). `verifyPrimaryKey`, `getEncryptionKeyPacket` etc instead only check the expiration time based on the primary user. So.. are we sure that checking every user is important?

Reply:
I think it's best to just be correct and return the expected result here for both this function and Key.prototype.getExpirationTime, because at least getExpirationTime is pretty frequently used by people who use the library. Plus it's debatable whether we are saving any time doing it this way vs. making it a Promise.all.